src/pybind/mgr/volumes/volume.py

```diff
@@ -0,0 +1,151 @@
+        spec.name = fs_name
+        try:
+            completion = self.mgr.add_stateless_service("mds", spec)
+            self.mgr._orchestrator_wait([completion])
+            orchestrator.raise_if_exception(completion)
+        except (ImportError, orchestrator.OrchestratorError):
+            return 0, "", "Volume created successfully (no MDS daemons created)"
+        except Exception as e:
+            # Don't let detailed orchestrator exceptions (python backtraces)
+            # bubble out to the user
+            return -errno.EINVAL, "", str(e)
+        return 0, "", ""
+
+    def teardown_mds(self, fs_name):
+        command = {'prefix': 'fs set', 'fs_name': fs_name, 'var': 'cluster_down', 'val': 'true'}
+        r, outb, outs = self.mgr.mon_command(command)
+        if r != 0:
+            return r, outb, outs
+        for mds in self.get_mds_names(fs_name):
+            command = {'prefix': 'mds fail', 'role_or_gid': mds}
+            r, outb, outs = self.mgr.mon_command(command)
+            if r != 0:
+                return r, outb, outs
+        return 0, "", ""
+
+    ### volume operations -- create, rm, ls
+
+    def create_volume(self, volname, size=None):
+        """
+        """
+        metadata_pool, data_pool = self.gen_pool_names(volname)
+        # create pools
+        r, outs, outb = self.create_pool(metadata_pool, 16, pg_num_min=16, pg_autoscale_factor=4.0)
+        if r != 0:
+            return r, outb, outs
+        r, outb, outs = self.create_pool(data_pool, 8)
+        if r != 0:
+            return r, outb, outs
+        # create filesystem
+        r, outb, outs = self.create_filesystem(volname, metadata_pool, data_pool)
+        if r != 0:
+            return r, outb, outs
+        # create mds
+        return self.create_mds(volname)
+
+    def delete_volume(self, volname):
+        """
+        """
+        # Tear down MDS daemons
+        try:
+            completion = self.mgr.remove_stateless_service("mds", volname)
+            self.mgr._orchestrator_wait([completion])
+            orchestrator.raise_if_exception(completion)
+        except (ImportError, orchestrator.OrchestratorError):
+            pass
+        except Exception as e:
+            # Don't let detailed orchestrator exceptions (python backtraces)
+            # bubble out to the user
+            return -errno.EINVAL, "", str(e)
+
+        # In case orchestrator didn't tear down MDS daemons cleanly, or
+        # there was no orchestrator, we force the daemons down.
+        if self.volume_exists(volname):
+            r, outb, outs = self.teardown_mds(volname)
+            if r != 0:
+                return r, outb, outs
+            r, outb, outs = self.remove_filesystem(volname)
+            if r != 0:
+                return r, outb, outs
+        else:
+            log.warning("Filesystem already gone for volume '{0}'".format(volname))
+            # log here...
+            pass
```

Comment:
Is pass required here?

Reply:
leftover from debug.. thanks for catching.